File: project_euler/Problem038/Problem038.py
from typing import List

class Problem038:
    def solution(self) -> None:
        digits = [False for _ in range(9)]
        for i in range(9487, 9233, -1):
            digits[:] = [False for _ in range(9)]
            if self.__is_pandigital(i, digits) and self.__is_pandigital(i << 1, digits):
                print('%d%d' % (i, i << 1))
                return

    def __is_pandigital(self, value: int, digits: List[bool]) -> bool:
        while value:
            index = value % 10 - 1
            if index < 0 or digits[index]:
                 return False
            digits[index] = True
            value //= 10
        return True

File: project_euler/Problem038/Problem038.cpp
#include <iostream>
#include <vector>
using namespace std;

class Problem038 {
public:
    void solution() {
        vector<bool> digits(9);
        for (int i = 9487; i >= 9234; --i) {
            digits.assign(9, false);
            if (is_pandigital(i, digits) && is_pandigital(i << 1, digits)) {
                cout << i << (i << 1) << endl;
                return;
            }
        }
    }

private:
    bool is_pandigital(int value, vector<bool> &digits) {
        while (value) {
            int index = value % 10 - 1;
            if (index < 0 || digits[index]) {
                return false;
            }
            digits[index] = true;
            value /= 10;
        }
        return true;
    }
};
